lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -103,6 +110,41 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
 //                        Canonicalize Memory Operations
 //===----------------------------------------------------------------------===//
 
+// Replace all uses of an original struct or tuple extract instruction, with the
+// given load instruction. The caller ensures that the load only loads the
+// extracted field.
+static void replaceUsesOfExtract(SingleValueInstruction *extract,
+                                 LoadInst *loadInst,
+                                 CanonicalizeInstruction &pass) {
+  assert(extract->getType() == loadInst->getType());
+
+  SingleValueInstruction *loadedVal = loadInst;
+  if (loadInst->getOwnershipQualifier() == LoadOwnershipQualifier::Copy) {
+    // Borrow the load-copied subelement, with precisely the same scope as
+    // the aggregate borrow.
+    assert(extract->getNumOperands() == 1);
+    auto *origBorrow = cast<BeginBorrowInst>(extract->getOperand(0));
+    auto *newBorrow = SILBuilderWithScope(origBorrow)
+                          .createBeginBorrow(loadInst->getLoc(), loadInst);
+    pass.notifyNewInstruction(newBorrow);
+
+    for (Operand *use : origBorrow->getUses()) {
```

Comment:
Do you think this is general enough to make it a helper on begin borrow. I do think that at some point we are going to create a helper to find the end borrows of a borrow scope. Just not sure if find singular end borrow given single use is useful enough.

But it does feel like this type of operation may be common when writing peepholes.

Reply:
Ok. I added this because we'll need it for sure.